source/extensions/filters/common/lua/protobuf_converter.cc

```diff
@@ -0,0 +1,214 @@
+      ProtobufConverterUtils::messageToLuaTable(state_,
+                                                reflection->GetMessage(*entry, value_field));
+      break;
+    case Protobuf::FieldDescriptor::CPPTYPE_INT32:
+    case Protobuf::FieldDescriptor::CPPTYPE_INT64:
+      lua_pushnumber(state_, reflection->GetInt64(*entry, value_field));
+      break;
+    case Protobuf::FieldDescriptor::CPPTYPE_BOOL:
+      lua_pushboolean(state_, reflection->GetBool(*entry, value_field));
+      break;
+    default:
+      lua_pushnil(state_);
+      break;
+    }
+  }
+
+  lua_State* state_;
+  const Protobuf::ReflectableMessage& message_;
+  const Protobuf::FieldDescriptor* field_;
+};
+
+// Helper class to handle repeated field conversion
+class RepeatedFieldConverter {
+public:
+  RepeatedFieldConverter(lua_State* state, const Protobuf::ReflectableMessage& message,
+                         const Protobuf::FieldDescriptor* field)
+      : state_(state), message_(message), field_(field) {}
+
+  void convert() {
+    lua_newtable(state_);
+    const auto* reflection = message_->GetReflection();
+    const int count = reflection->FieldSize(*message_, field_);
+
+    for (int i = 0; i < count; i++) {
+      lua_pushinteger(state_, i + 1); // Lua uses 1-based indexing
+      pushRepeatedValue(reflection, i);
+      lua_settable(state_, -3);
+    }
+  }
+
+private:
+  void pushRepeatedValue(const Protobuf::Reflection* reflection, int index) {
+    switch (field_->cpp_type()) {
+    case Protobuf::FieldDescriptor::CPPTYPE_STRING:
+      pushLuaString(state_, reflection->GetRepeatedString(*message_, field_, index));
+      break;
+    case Protobuf::FieldDescriptor::CPPTYPE_MESSAGE:
+      ProtobufConverterUtils::messageToLuaTable(
+          state_, reflection->GetRepeatedMessage(*message_, field_, index));
+      break;
+    case Protobuf::FieldDescriptor::CPPTYPE_INT32:
+    case Protobuf::FieldDescriptor::CPPTYPE_INT64:
+      lua_pushnumber(state_, reflection->GetRepeatedInt64(*message_, field_, index));
+      break;
+    case Protobuf::FieldDescriptor::CPPTYPE_BOOL:
+      lua_pushboolean(state_, reflection->GetRepeatedBool(*message_, field_, index));
+      break;
+    default:
+      lua_pushnil(state_);
+      break;
+    }
+  }
+
+  lua_State* state_;
+  const Protobuf::ReflectableMessage& message_;
+  const Protobuf::FieldDescriptor* field_;
+};
+
+} // namespace
+
+void ProtobufConverterUtils::messageToLuaTable(lua_State* state, const Protobuf::Message& message) {
+  lua_newtable(state);
+
+  Protobuf::ReflectableMessage reflectable_message = createReflectableMessage(message);
+  const auto* reflection = reflectable_message->GetReflection();
+
+  std::vector<const Protobuf::FieldDescriptor*> fields;
+  reflection->ListFields(*reflectable_message, &fields);
+
+  for (const auto* field : fields) {
+    pushLuaString(state, field->name());
+
+    if (field->is_map()) {
+      MapFieldConverter(state, reflectable_message, field).convert();
```

Comment:
This API is a little strange, where you always construct a class and then call convert on it. I'm not sure it needs to be a class at all; can you change it into just a function? Or alternately, expose a static method `convert` that takes all the constructor parameters, and make the constructor private.